src/Sylius/Bundle/CoreBundle/Form/Type/PaymentMethodType.php

```diff
@@ -0,0 +1,219 @@
+    {
+        parent::__construct($paymentMethodClass, $validationGroups);
+        $this->registry = $registry;
+        $this->gatewayConfigStore = $gatewayConfigStore;
+        $this->defaultGateways = $defaultGateways;
+        $this->translator = $translator;
+
+        $this->initializeGatewayConfigs();
+    }
+
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        parent::buildForm($builder, $options);
+
+        $this->buildGatewayConfigForm($builder);
+
+        $builder->addEventListener(FormEvents::POST_SUBMIT, array($this, 'processGatewayConfig'));
+    }
+
+    /**
+     * @param FormBuilderInterface $builder
+     */
+    protected function buildGatewayConfigForm(FormBuilderInterface $builder)
+    {
+        foreach ($this->gatewayConfigs as $factoryName => $config) {
+
+            /**
+             * @var  $gatewayConfig \Sylius\Bundle\PayumBundle\Model\GatewayConfig
+             */
+            $gatewayConfig = null;
+
+            if ($gatewayConfigs = $this->gatewayConfigStore->findBy(array('gatewayName' => $factoryName))) {
+                $gatewayConfig = array_shift($gatewayConfigs);
+            }
+
+            $configForm = $builder->create($factoryName, FormType::class, array(
+                'mapped' => false,
+                'data' => $gatewayConfig ? $gatewayConfig->getConfig() : null
+            ));
+
+            foreach ($config['payum.default_options'] as $name => $value) {
+                $type = is_bool($value) ? CheckboxType::class : TextType::class;
+
+                $options = array(
+                    'required' => in_array($name, $config['payum.required_options'])
+                );
+
+                $configForm->add($name, $type, $options);
+            }
+
+            $builder->add($configForm);
+        }
+    }
+
+
+    /**
+     * @param FormEvent $event
+     */
+    public function processGatewayConfig(FormEvent $event)
+    {
+        /**
+         * @var  $paymentMethod \Sylius\Component\Payment\Model\PaymentMethod
+         */
+        $paymentMethod = $event->getData();
+        $form = $event->getForm();
+
+        $factoryName = $paymentMethod->getGateway();
+        $configForm = $form->has($factoryName) ? $form->get($factoryName) : null;
+
+        if (!$configForm) {
+            return;
+        }
+
+        $data = $configForm->getData();
+
+        $config = $this->gatewayConfigs[$factoryName];
+
+        if (!isset($config['payum.required_options']))
+            return;
+
+        foreach ($config['payum.required_options'] as $option) {
+            if (!isset($data[$option]) || empty($data[$option])) {
+                $formError = new FormError($this->translator->trans('This value should not be blank.'));
```

Comment:
Why did you hardcode this validation? It should be placed in a validation.xml file.